data-azure-cosmos/src/main/java/io/micronaut/data/cosmos/operations/DefaultCosmosRepositoryOperations.java

```diff
@@ -253,12 +261,144 @@ public Optional<Number> executeUpdate(PreparedQuery<?, Number> preparedQuery) {
+            while (iterator.hasNext()) {
+                T entity = iterator.next();
+                ObjectNode tree = serializeToTree(entity, arg);
+                CosmosItemResponse<?> cosmosItemResponse = container.deleteItem(tree, options);
+                if (cosmosItemResponse.getStatusCode() == HttpResponseStatus.NO_CONTENT.code()) {
+                    deletedCount++;
+                }
+            }
+        } else {
+            List<ObjectNode> items = new ArrayList<>();
+            while (iterator.hasNext()) {
+                T entity = iterator.next();
+                items.add(serializeToTree(entity, arg));
+            }
+            deletedCount = bulkDelete(container, items.iterator(), persistentEntity, Optional.empty());
+        }
+        return Optional.of(deletedCount);
+    }
+
+    @Override
+    public Optional<Number> executeDelete(PreparedQuery<?, Number> preparedQuery) {
+        boolean isRawQuery = preparedQuery.getAnnotationMetadata().stringValue(Query.class, DataMethod.META_MEMBER_RAW_QUERY).isPresent();
+        if (isRawQuery) {
+            throw new IllegalStateException("Cosmos Db does not support raw delete queries.");
+        }
+        RuntimePersistentEntity<?> persistentEntity = runtimeEntityRegistry.getEntity(preparedQuery.getRootEntity());
+        CosmosContainer container = getContainer(persistentEntity);
+        List<SqlParameter> paramList = bindParameters(preparedQuery);
+        SqlQuerySpec querySpec = new SqlQuerySpec("SELECT * " + preparedQuery.getQuery(), paramList);
+        CosmosQueryRequestOptions queryRequestOptions = new CosmosQueryRequestOptions();
+        Optional<PartitionKey> optPartitionKey = preparedQuery.getParameterInRole(Constants.PARTITION_KEY_ROLE, PartitionKey.class);
+        optPartitionKey.ifPresent(queryRequestOptions::setPartitionKey);
+        CosmosPagedIterable<ObjectNode> result = container.queryItems(querySpec, queryRequestOptions, ObjectNode.class);
+        Iterator<ObjectNode> iterator = result.iterator();
+        String partitionKeyDefinition = getPartitionKey(persistentEntity);
+        int deletedCount;
+        if (optPartitionKey.isPresent()) {
+            // Do bulk delete if there is partition key provided
+            deletedCount = bulkDelete(container, iterator, persistentEntity, optPartitionKey);
+        } else if (StringUtils.isNotEmpty(partitionKeyDefinition)) {
+            // Bulk delete if there is partition key defined on the entity, so we can get its value
+            deletedCount = bulkDelete(container, iterator, persistentEntity, Optional.empty());
+        } else {
+            // If no partition key provided or defined on the entity (should not happen), delete one by one since partition key is needed for bulk delete
+            deletedCount = 0;
+            CosmosItemRequestOptions options = new CosmosItemRequestOptions();
+            while (iterator.hasNext()) {
```

Comment:
Also here

Reply:
There is different, we get iterator from query execution result 
```
CosmosPagedIterable<ObjectNode> result = container.queryItems(querySpec, queryRequestOptions, ObjectNode.class);
Iterator<ObjectNode> iterator = result.iterator();
```

Reply:
It should be possible to write `for(ObjectNode :  result) {` because `for each` works with anything that implements iterable.

Reply:
Ok, changed.